Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+      }
+
+      // binary file version 2 now contains the unix time stamp
+      if (binaryFileVersion_ == 2) {
+        uint16_t unixtime[4];
+        currentFile_->read((char *) &unixtime[0], sizeof(unixtime));
+        uint64_t time = ((uint64_t) unixtime[3] << 48) | ((uint64_t) unixtime[2] << 32) | ((uint64_t) unixtime[1] << 16) | ((uint64_t) unixtime[0]);
+        if (verbosityLevel_ > 0) {
+          using time_point = std::chrono::system_clock::time_point;
+          time_point header_timepoint(std::chrono::duration_cast<time_point::duration>(std::chrono::milliseconds(time)));
+          std::time_t header_t = std::chrono::system_clock::to_time_t(header_timepoint);
+          std::cout << "Unix Time: " << std::put_time(std::gmtime(&header_t), "%c %Z") << std::endl;
+        }
+      }
+
+      // Get event information for this trigger and put in the event
+      std::unique_ptr<mu2e::STMTestBeamEventInfo> outputEvtInfo(new mu2e::STMTestBeamEventInfo(trigger_header[0].getTriggerMode(), trigger_header[0].getTriggerTime()));
+      art::put_product_in_principal(std::move(outputEvtInfo), *outE, myModuleLabel_);
+
+      // Get the number of slices in this trigger
+      int n_slices = trigger_header[0].getNSlices();
+      for (int i_slice = 0; i_slice < n_slices; ++i_slice) {
+        STMTestBeam::SliceHeader slice_header[1];
+        // Read the slice header
+        if(currentFile_->read((char *) &slice_header[0], sizeof(STMTestBeam::SliceHeader))) {
+          if(verbosityLevel_ > 0) {
+            std::cout << slice_header[0] << std::endl;
+          }
+
+          // Get the number of ADC samples in this slice
+          unsigned long int n_adc_samples = slice_header[0].getNADC();
+
+          // Read the ADC samples into a vector for later
+          std::vector<int16_t> adcs;
+          adcs.reserve(n_adc_samples);
+          int16_t adc[1];
+          for (unsigned long int i_adc_sample = 0; i_adc_sample < n_adc_samples; ++i_adc_sample) {
+            currentFile_->read((char *) &adc[0], sizeof(int16_t));
+            adcs.push_back(adc[0]);
+          }
+
+          // Create the STMWaveformDigi and put it in the event
+          STMWaveformDigi stm_waveform(trigger_header[0].getTriggerOffset(), adcs);
+          if (channel_ == STMChannel::HPGe) {
+            outputHPGeWaveformDigis->push_back(stm_waveform);
+          }
+          else if (channel_ == STMChannel::LaBr) {
+            outputLaBrWaveformDigis->push_back(stm_waveform);
+          }
+          else {
+            throw cet::exception("FromSTMTestBeamData") << "Trying to create a waveform with an invalid STMChannel (" << channel_ << ")" << std::endl;
+          }
+        }
+        else { return false; }
+      }
+    }
+    else { return false; }
+
+    art::put_product_in_principal(std::move(outputHPGeWaveformDigis), *outE, myModuleLabel_, "HPGe");
```

Comment:
Hi Rob, I think downstream we want to assume that both HPGe and LaBr collections are in the event (even if they are empty). This should allow us to have one fcl file for processing so that if one detector is physically off that fcl file will still run